Writing on a screen

tpltnt opened the pull request:
Hi there,

I wrote a tutorial for displaying text on a screen using different fonts.

Cheers,
tpltnt

tpltnt committed:
added text demo source

tpltnt committed:
text added

tpltnt committed:
added tutorial to summary / ToC

tpltnt committed:
make copied code compile

tpltnt commented:
ping

mitchmindtree commented:
Thanks a lot for this @tpltnt! I haven't had a chance to properly read through and review this just yet but I hope to soon.

I just had a glance and the first suggestion that comes to mind is that perhaps it might be nice to include an image after each step so that the user can see what is expected to be on screen?

tpltnt committed:
added screenshots

tpltnt committed:
added screenshots to text

tpltnt commented:
I added three screenshots.

tpltnt commented:
ping

tpltnt commented:
ping

tpltnt commented:
What is blocks this PR from merging?

mitchmindtree commented on src/SUMMARY.md:
```diff
@@ -12,6 +12,7 @@
     - [Upgrading to a New Release](./getting_started/upgrading.md)
 - [Tutorials](./tutorials.md)
     - [Basics - Anatomy of a Nannou App](./tutorials/basics/anatomy-of-a-nannou-app.md)
+    - [Basics - Writing on the Screen](./tutorials/basics/writing-on-the-screen.md)
```
Do you mind if we rename this to "GUI - Displaying Text"?

tpltnt replied:
It looses the pun, but I can do that. Also: is "GUI" considered "basic"?

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+# Writing on the Screen
```
Would you mind moving this into a new `src/tutorials/gui/displaying-text.md` file?

tpltnt replied:
Done that.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
````diff
@@ -0,0 +1,767 @@
+all the functions we need.
+```rust,no_run
+#![allow(unused_imports)]
+extern crate nannou;  // tell rust that we want to use nannou
````
As this is only required by the nannou-guide tests, I think we can hide this line from this section. You can hide a line from the generated documentation by putting a `#` in front of it like so:
```
# #![allow(unused_imports)]
# extern crate nannou;
```

tpltnt replied:
done that

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+# }
+#
+// handle events and update the (world) model accordingly
+fn event(_app: &App, model: &mut Model, event: WindowEvent) {
```
Rather than using an `event` function, the code might look a lot neater using a [`key_presed` function](https://github.com/nannou-org/nannou/blob/master/examples/all_functions.rs#L87).

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
````diff
@@ -0,0 +1,767 @@
+fn view(app: &App, model: &Model, frame: Frame) -> Frame {
+    frame
+}
+```
````
It looks like the intention is to start with the application skeleton and then take a closer look at each section step by step. In that case, perhaps the prelude import section above should come after this section?

tpltnt replied:
This may be one approach. I prefer not going back and rework code I already went past (in terms of explanation).

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+    if let Err(e) = res {
+        println!("{}", e);
+        panic!("loading font failed");
+    }
```
This can be expressed in a more elegant manner:

```rust
font_map
    .insert_from_file("assets/fonts/retroscape/Retroscape.ttf")
    .expect("loading retroscape font failed");
```
This will `panic!` with the specified message followed by the error message if an error occurs.

mitchmindtree replied:
Please also update the other code snippets for this change.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+    if let Err(e) = res {
+        println!("{}", e);
+        panic!("loading font failed");
+    }
```
Same as the comment above.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
````diff
@@ -0,0 +1,767 @@
+# fn view(app: &App, model: &Model, frame: Frame) -> Frame {
+#    frame
+# }
+```
````
You should be able to get rid of a lot of the hidden code from this code snippet as only the `main` function is being demonstrated here. The same goes for the following sections that also unnecessarily include large amounts of hidden code.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+It contains the field `ui` to hold the [user interface](https://en.wikipedia.org/wiki/User_interface)
+(actually the [GUI](https://en.wikipedia.org/wiki/Graphical_user_interface)). The UI contains
+[widgets](https://en.wikipedia.org/wiki/Widget_(GUI)) to make it interactive. We use the `Ids`
+struct as value for for die field `widget_ids`. This struct only contains one field: `text`.
```
Looks like there was a typo here

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+(actually the [GUI](https://en.wikipedia.org/wiki/Graphical_user_interface)). The UI contains
+[widgets](https://en.wikipedia.org/wiki/Widget_(GUI)) to make it interactive. We use the `Ids`
+struct as value for for die field `widget_ids`. This struct only contains one field: `text`.
+That field contains the widget ID for a text field (to display on screen). The text field is
```
I don't think these parentheses are necessary, I think this should be fine?

```
This field contains the widget ID for a text field.
```

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+a widget, but not interactive in the traditional sense in that you can point and click it.
+Using an extra struct for the widget IDs keeps them all in one place and allows consistent
+access like `model.widget_ids.text`. The message we want to display on the screen is stored
+in the `message` field of the model (which holds a `String`). Text on a computer screen is
```
Unnecessary parentheses

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+in the `message` field of the model (which holds a `String`). Text on a computer screen is
+rendered using a [font](https://en.wikipedia.org/wiki/Computer_font). Usually there are many
+fonts to choose from on a computer. Nannou uses an internal registry called a "font map"
+to keep track of them. A font in the font map has a (numerical) ID to uniquely identify it.
```
Unnecessary parentheses

tpltnt replied:
Which do you mean exactly?

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+we make an `Ids` struct and generate a widget ID for the text field. `ui.generate_widget_id()`
+is necessary to gurantee unique IDs per widget so each can interact without internal confusion.
+
+Now that the UI widgets are taken care of the can handle the fonts to display the message text.
```
This sentence does not make sense - maybe a typo?

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+is necessary to gurantee unique IDs per widget so each can interact without internal confusion.
+
+Now that the UI widgets are taken care of the can handle the fonts to display the message text.
+We learned already that Nannou tracks fonts in a (global) font map. We get access to it by
```
I think `(global)` can be removed, as global has a different meaning in Rust. The font map is owned by the `Ui` and only lives as long as the `Ui` does.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+font. But wait? We know that the font is a file ... where is that loaded from? You are right
+in that fonts are files. They come in different formats (like [TrueType](https://en.wikipedia.org/wiki/TrueType),
+or [OpenType](https://en.wikipedia.org/wiki/OpenType)). These fonts are assets for the app.
+Assets are stored in the `assets/` directory. The default for is [Noto (sans regular)](https://www.google.com/get/noto/)
```
> The default for is

Typo?

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+Assets are stored in the `assets/` directory. The default for is [Noto (sans regular)](https://www.google.com/get/noto/)
+and it needs to be placed in the `DEFAULT_FONT_PATH`, i.e. "fonts/NotoSans/NotoSans-Regular.ttf".
+We also want different fonts / typefaces. For this tutorial the fonts [retroscape](https://fontlibrary.org/en/font/retroscape)
+and [voynich](https://fontlibrary.org/en/font/voynich-font) are also used. They need to be
```
Neither of these URLs are working for me at the moment. Maybe the site is down? Can you confirm that the license for these fonts is permissive enough for us to use them in this tutorial?

tpltnt replied:
Maybe. They were working at the time of writing and can still be found in the caches of search engines. Retroscape is Creative Commons CC0, voynich was permissive. I will replace them since the website is down.

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+downloaded and placed in the "assets/fonts/" directory. The font "retroscape" is loaded
+into the font map using the function `insert_from_file()`. This may go wrong, so some
+error checking should be done. Since the names of the variables match the names of the
+fields in the model, so they can be placed there directly.
```
The word "so" is unnecessary here

tpltnt replied:
Which "so"?

mitchmindtree commented on src/tutorials/basics/writing-on-the-screen.md:
```diff
@@ -0,0 +1,767 @@
+
+The model is set up now, let's look at the remaining logic of the app.
+
+First we handle all (window) events.
```
Unnecessary parentheses

mitchmindtree reviewed (changes requested):
Thanks again @tpltnt and sorry again for taking a while to address this. It is such a large PR and it has been difficult finding time to review as there are many things that I think need to be addressed.

A side-note on parentheses: In general it is good to avoid them when writing unless necessary as they break the flow of reading. Often a sentence can be re-structured to avoid them, and often they are unnecessary if the sentence reads fine without them.

I have reviewed about two thirds of this but am out of time for now - please read over your writing to make sure there are no typos or spelling mistakes. There are quite a few errors in here that lead me to suspect that you have not read this after writing it. Proofreading will help to speed up the PR process.

I've left more detailed comments throughout the code.

tpltnt committed:
changed tutorial name to 'displaying text'

tpltnt committed:
removed some code from rendering

tpltnt committed:
replaced voynich font with qwigley

tpltnt committed:
fixed typo and clarified sentence

tpltnt committed:
fixed typo

tpltnt committed:
added explanation of more complicated event processing

tpltnt committed:
simplified font loading

tpltnt committed:
typo fix

tpltnt committed:
removed parentheses

tpltnt committed:
updated screenshots

tpltnt committed:
updated image descriptions

tpltnt committed:
reworked text

tpltnt committed:
fixed URL

tpltnt committed:
fixed code comment

tpltnt commented:
Here is an improved version.

tpltnt commented:
running `cd book-tests && cargo test -v` locally yields
```
running 0 tests

test result: ok. 21 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```
Travis seems to be off (given that the tests for "anatomy of a nannou app" also fail)

tpltnt committed:
allowing unused variables

tpltnt committed:
removed unnecessary code

tpltnt committed:
leveraging linter groups

tpltnt committed:
port to nannou 0.12.0

tpltnt commented:
All requested changes are addressed and the code is ported to 0.12.0 ... can we merge it now?

tpltnt commented:
ping

tpltnt commented:
ping

JoshuaBatty commented:
Hi @tpltnt thanks for putting this tutorial together, we really appreciate the input. But we have recently made drawing text to a window much simpler in nannou. See the [simple_text example](https://github.com/nannou-org/nannou/blob/master/examples/simple_text.rs). 

As such, I don't think we should include this approach in the guide as it is overly verbose compared to the new method which has been designed to be as simple as possible.

tpltnt commented:
A hint that font things are (going to be) reworked or any reaction way more than half a year ago would have been helpful. I tried to work with what I was able to discern from the documentation back then. I addressed all issues raised and even ported the code to a new version. All these efforts are basically wasted now.

mitchmindtree commented:
My apologies @tpltnt, you're right we should have kept you up to date w.r.t. the text API changes and linked you to the existing issues at nannou. Unfortunately as nannou is still very young, it will frequently be the case where past work is made redundant by new improvement or nicer APIs - it's not always easy to foresee what these changes will be. That said, I'll try harder to improve communications in the future!

tpltnt commented:
@mitchmindtree that already was good communication :)
I can imagine that things fall through the cracks as this project is still young and has so many evolving parts.
I am frustrated because of the wasted effort. I provided a tutorial as it was asked for in the guide. I addressed all issues raised in the first review and was left in the dark afterwards.
My advice: Either merge after all issues have been addressed or give people a heads-up about potential changes. (Maybe topic-related tags on issues help)
I observe that there are only two people in the nannou github project. This smells like a bottleneck to me because the two of you take care of so many things. Maybe allow other people access on sub-projects? The guide for example is less critical than the main codebase if something goes wrong. Maybe think about delegating some management (work) to trustworthy people in the community, thus reducing the number of things you have to juggle while also allowing quicker responses and increasing resilience.
